File: src/main/java/io/github/emcw/map/api/Nations.java
package io.github.emcw.map.api;

import io.github.emcw.caching.BaseCache;
import io.github.emcw.caching.CacheOptions;

import io.github.emcw.map.entities.Nation;
import io.github.emcw.exceptions.MissingEntryException;
import io.github.emcw.interfaces.ILocatable;

import org.jetbrains.annotations.NotNull;

import java.util.Map;

public class Nations extends BaseCache<Nation> implements ILocatable<Nation> {

    public Nations(CacheOptions options) {
        super(options);

        setUpdater(this::forceUpdate);
        build();
    }

    public void tryUpdate() {
        tryExpire();
        updateCache(false);
    }

    public void forceUpdate() {
        updateCache(true);
    }

    private void updateCache(Boolean force) {
//        if (!empty() && !force) return;
//
//        // Parse map data into usable Nation objects.
//        SquaremapParser.parseMapData(parent.getMapName(), true, true, false);
//        Cache<String, Nation> nations = SquaremapParser.parsedNations(parent.getMapName());
//
//        if (!nations.asMap().isEmpty())
//            setCache(nations);
    }

    public Map<String, Nation> all() {
        tryUpdate();
        return super.all();
    }

    @Override
    public Nation single(String name) throws MissingEntryException {
        tryUpdate();
        return super.single(name);
    }

    public Map<String, Nation> get(String @NotNull ... keys) {
        tryUpdate();
        return super.get(keys);
    }
}

File: src/main/kotlin/io/github/emcw/map/api/Nations.kt
package io.github.emcw.map.api

import io.github.emcw.EMCMap
import io.github.emcw.caching.BaseCache
import io.github.emcw.caching.CacheOptions
import io.github.emcw.exceptions.MissingEntryException
import io.github.emcw.interfaces.ILocatable
import io.github.emcw.map.entities.Nation
import io.github.emcw.utils.DataParser.parseMapData
import io.github.emcw.utils.DataParser.parsedNations

class Nations(private val parent: EMCMap, options: CacheOptions) : BaseCache<Nation?>(options), ILocatable<Nation?> {
    init {
        setUpdater { forceUpdate() }
        build()
    }

    fun tryUpdate() {
        tryExpire()
        updateCache(false)
    }

    fun forceUpdate() {
        updateCache(true)
    }

    private fun updateCache(force: Boolean) {
        if (!empty() && !force) return

        // Parse map data into usable Nation objects.
        parseMapData(parent.mapName, true, true, false)
        val nations = parsedNations(
            parent.mapName
        )
        if (!nations.asMap().isEmpty()) setCache(nations)
    }

    override fun all(): Map<String?, Nation?>? {
        tryUpdate()
        return super.all()
    }

    @Throws(MissingEntryException::class)
    override fun single(name: String?): Nation? {
        tryUpdate()
        return super.single(name)
    }

    override fun get(vararg keys: String): Map<String?, Nation?>? {
        tryUpdate()
        return super.get(*keys)
    }
}